packages/base/index.d.ts

```diff
@@ -0,0 +1,356 @@
+declare namespace folktale {
+  // -- Core
+  interface Lambda {
+    compose<A, B, C>(f: (_: B) => C, g: (_: A) => B): (_: A) => C;
+    constant<A, B>(value: A): (_: B) => A;
+    identity<A>(value: A): A;
+  }
+
+  interface Object {
+    fromPairs<A>(pairs: [string, A][]): { [key: string]: A};
+    toPairs<A>(object: {[key: string]: A}): [string, A][];
+    values<A>(object: {[key: string]: A}): A[];
+    mapValues<A, B>(object: {[key: string]: A}, fn: (_: A) => B): {[key: string]: B};
+    mapEntries: MapEntries
+  }
+
+  type Dict<A> = { [key: string]: A };
+
+  interface MapEntries {
+    <A, B>(object: Dict<A>, transform: (pair: [string, A]) => [string, B], define: (object: Dict<B>, key: string, value: B) => Dict<B>): Dict<B>
+
+    overwrite<A, B>(object: Dict<A>, transform: (pair: [string, A]) => [string, B]): Dict<B>
+    unique<A, B>(object: Dict<A>, transform: (pair: [string, A]) => [string, B]): Dict<B>
+  }
+
+  interface Core {
+    lambda: Lambda;
+    object: Object;
+  }
+
+  // -- Maybe
+  type Semigroup<A> = A & {
+    concat(this: Semigroup<A>, that: Semigroup<A>): Semigroup<A>
+  }
+
+  type Maybe<A> = Just<A> | Nothing<A>
+
+  interface Just<A> {
+    __type: A;
+    __tag: 'Just';
+    value: A;
+    map<B>(f: (_: A) => B): Maybe<B>;
+    apply<B>(this: Maybe<(_: A) => B>, f: Maybe<A>): Maybe<B>;
```

Comment:
this type incorrect i think. We use the Generic variable `A` here for other Maybe, it cause typescript infer the `this` context must be `Maybe<(_: (_: A) => B) => B>`. To make it clear try: `ma.map(tuple).apply(mb)`.

Reply:
`ap<B, C>(this: Maybe<(_: B) => C>, other: Task<B>): Task<C>` might do the trick, don't use the `A` variable in Generic.

Reply:
Yeah, the A type variable should be local to that method.